File: dotnet/Vfx/OpenStack.Vfx/Vfx.cs
using Microsoft.Extensions.FileSystemGlobbing;
using OpenStack.Vfx.Disc;
using OpenStack.Vfx.N64;
using OpenStack.Vfx.X3ds;
using SharpCompress.Archives.SevenZip;
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;

[assembly: InternalsVisibleTo("OpenStack.Vfx.Program")]

namespace OpenStack.Vfx;

/// <summary>
/// VfxExtensions
/// </summary>
public static class VfxExtensions {
}

#region FileSystem

/// <summary>
/// FileSystem
/// </summary>
public abstract class FileSystem {
    public abstract IEnumerable<string> Glob(string path, string searchPattern);
    public abstract bool FileExists(string path);
    public abstract (string path, long length) FileInfo(string path);
    public abstract Stream Open(string path, string mode = null);
    public virtual FileSystem Next() => this;
    public IEnumerable<string> FindPaths(string path, string searchPattern) {
        // expand
        int expandStartIdx, expandMidIdx, expandEndIdx;
        if ((expandStartIdx = searchPattern.IndexOf('(')) != -1 &&
            (expandMidIdx = searchPattern.IndexOf(':', expandStartIdx)) != -1 &&
            (expandEndIdx = searchPattern.IndexOf(')', expandMidIdx)) != -1 &&
            expandStartIdx < expandEndIdx) {
            foreach (var expand in searchPattern.Substring(expandStartIdx + 1, expandEndIdx - expandStartIdx - 1).Split(':'))
                foreach (var found in FindPaths(path, searchPattern.Remove(expandStartIdx, expandEndIdx - expandStartIdx + 1).Insert(expandStartIdx, expand)))
                    yield return found;
            yield break;
        }
        foreach (var file in Glob(path, searchPattern)) yield return file;
    }

    /// <summary>
    /// TryAdvance
    /// </summary>
    /// <param name="basePath"></param>
    /// <param name="path"></param>
    /// <returns></returns>
    public FileSystem Advance(string basePath, string path)
         => Path.GetExtension(path).ToLowerInvariant() switch {
             ".zip" => new ZipFileSystem(this, path, basePath),
             ".7z" => new SevenZipFileSystem(this, path, basePath),
             //".iso" => new IsoFileSystem(this, path, basePath),
             ".bin" or ".cue" => new DiscFileSystem(this, Glob("", "*.cue").Single(), basePath),
             ".n64" or ".v64" or ".z64" => new N64FileSystem(this, path, basePath),
             ".3ds" => new X3dsFileSystem(this, path, basePath),
             _ => null
         };

    protected FileSystem Next2(string basePath, int count, Func<string> firstFunc, Func<FileSystem> elseFunc) {
        if (count == 0) return this;
        var first = firstFunc() ?? "";
        return count == 1 || first.EndsWith(".bin", StringComparison.OrdinalIgnoreCase) || first.EndsWith(".cue", StringComparison.OrdinalIgnoreCase)
            ? Advance(basePath, first) ?? this
            : elseFunc();
    }

    /// <summary>
    /// Creates the matcher.
    /// </summary>
    /// <param name="searchPattern">The searchPattern.</param>
    /// <returns></returns>
    public static Func<string, bool> CreateMatcher(string searchPattern) {
        if (string.IsNullOrEmpty(searchPattern)) return x => true;
        var wildcardCount = searchPattern.Count(x => x.Equals('*'));
        if (wildcardCount <= 0) return x => x.Equals(searchPattern, StringComparison.CurrentCultureIgnoreCase);
        else if (wildcardCount == 1) {
            var newPattern = searchPattern.Replace("*", "");
            if (searchPattern.StartsWith("*")) return x => x.EndsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
            else if (searchPattern.EndsWith("*")) return x => x.StartsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
        }
        var regexPattern = $"^{Regex.Escape(searchPattern).Replace("\\*", ".*")}$";
        return x => {
            try { return Regex.IsMatch(x, regexPattern); }
            catch { return false; }
        };
    }
}

/// <summary>
/// AggregateFileSystem
/// </summary>
public class AggregateFileSystem(FileSystem[] aggreate) : FileSystem {
    readonly FileSystem[] aggreate = aggreate;

    public override IEnumerable<string> Glob(string path, string searchPattern) => aggreate.SelectMany(x => x.Glob(path, searchPattern));
    public override bool FileExists(string path) => aggreate.Any(x => x.FileExists(path));
    public override (string path, long length) FileInfo(string path) => aggreate.Select(x => x.FileInfo(path)).FirstOrDefault(x => x.path != null);
    public override Stream Open(string path, string mode) => aggreate.Select(x => x.Open(path, mode)).FirstOrDefault(x => x != null);
}

/// <summary>
/// VirtualFileSystem
/// </summary>
public class VirtualFileSystem(Dictionary<string, byte[]> virtuals) : FileSystem {
    readonly Dictionary<string, byte[]> Virtuals = virtuals;

    public override IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = CreateMatcher(searchPattern);
        return Virtuals.Keys.Where(matcher);
    }
    public override bool FileExists(string path) => Virtuals.ContainsKey(path);
    public override (string path, long length) FileInfo(string path) => Virtuals.TryGetValue(path, out var z) ? (path, z != null ? z.Length : 0) : (null, 0);
    public override Stream Open(string path, string mode) => Virtuals.TryGetValue(path, out var z) ? z != null ? new MemoryStream(z) : new MemoryStream() : null;
}

/// <summary>
/// DirectoryFileSystem
/// </summary>
public class DirectoryFileSystem(string baseRoot, string basePath) : FileSystem {
    readonly string BaseRoot = baseRoot; readonly string BasePath = basePath; string Root = baseRoot; int Skip = baseRoot.Length + 1;

    public override IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = new Matcher();
        matcher.AddIncludePatterns([string.IsNullOrEmpty(searchPattern) ? "**/*" : searchPattern]);
        return [.. matcher.GetResultsInFullPath(Path.Combine(Root, path)).Select(x => x[Skip..])];
    }
    public override bool FileExists(string path) => File.Exists(Path.Combine(Root, path));
    public override (string path, long length) FileInfo(string path) => File.Exists(path = Path.Combine(Root, path)) ? (path[Skip..], new FileInfo(Path.Combine(Root, path)).Length) : (null, 0);
    public override Stream Open(string path, string mode) => mode != "Height"
        ? File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Read, FileShare.Read)
        : File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Write, FileShare.Write);
    public override FileSystem Next() {
        if (File.Exists(Root) || Path.GetFileName(Root).Contains('*')) {
            Root = Path.GetDirectoryName(Root); Skip = Root.Length + 1;
            return Advance(BasePath, Path.GetFileName(BaseRoot))?.Next() ?? this;
        }
        return Next2(BasePath, -1, () => Directory.EnumerateFiles(Root).FirstOrDefault(), () => {
            if (!string.IsNullOrEmpty(BasePath)) { Root = Path.Combine(BaseRoot, BasePath); Skip = Root.Length + 1; }
            return this;
        });
    }
}

#endregion

#region FileSystem : Network

/// <summary>
/// NetworkFileSystem
/// </summary>
public class NetworkFileSystem : FileSystem {
    public NetworkFileSystem(Uri uri) {
        if (uri == null) throw new ArgumentNullException(nameof(uri));
        var pathOrPattern = uri.LocalPath;
        var searchPattern = Path.GetFileName(pathOrPattern);
        //var path = Path.GetDirectoryName(pathOrPattern);
        // file
        if (!string.IsNullOrEmpty(searchPattern)) throw new ArgumentOutOfRangeException(nameof(pathOrPattern), pathOrPattern); //: Web single file access to supported.

        //options = PakOption.Stream;
        //searchPattern = Path.GetFileName(path);
        //path = Path.GetDirectoryName(path);
        //if (path.Contains('*')) throw new NotSupportedException("Web wildcard folder access");
        //host = new UriBuilder(uri) { Path = $"{path}/", Fragment = null }.Uri;
        //if (searchPattern.Contains('*'))
        //{
        //    var set = new HttpHost(host).GetSetAsync().Result ?? throw new NotSupportedException(".set not found. Web wildcard access");
        //    var pattern = $"^{Regex.Escape(searchPattern.Replace('*', '%')).Replace("_", ".").Replace("%", ".*")}$";
        //    return set.Where(x => Regex.IsMatch(x, pattern)).ToArray();
        //}
        //return new[] { searchPattern };
    }
    public override IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = new Matcher();
        matcher.AddIncludePatterns([searchPattern]);
        return matcher.GetResultsInFullPath(searchPattern).ToList();
    }
    public override bool FileExists(string path) => File.Exists(path);
    public override (string path, long length) FileInfo(string path) => File.Exists(path) ? (path, 0) : (null, 0);
    public override Stream Open(string path, string mode) => null;
}

#endregion

#region FileSystem : Archive

/// <summary>
/// ZipFileSystem
/// </summary>
public class ZipFileSystem(FileSystem vfx, string path, string basePath) : FileSystem {
    readonly ZipArchive Arc = new(vfx.Open(path), ZipArchiveMode.Read);
    string Root = string.Empty;

    public override IEnumerable<string> Glob(string path, string searchPattern) {
        var root = Path.Combine(Root, path); var skip = root.Length;
        var matcher = CreateMatcher(searchPattern);
        return [.. Arc.Entries.Where(x =>
        {
            var fn = x.FullName;
            return !fn.EndsWith("/") && fn.Length > skip && fn.StartsWith(root) && matcher(fn[skip..]);
        }).Select(x => x.FullName[skip..])];
    }
    public override bool FileExists(string path) => Arc.GetEntry(Path.Combine(Root, path)) != null;
    public override (string path, long length) FileInfo(string path) { var x = Arc.GetEntry(Path.Combine(Root, path)); return x != null ? (x.Name, x.Length) : (null, 0); }
    public override Stream Open(string path, string mode) => Arc.GetEntry(Path.Combine(Root, path)).Open();
    public override FileSystem Next() => Next2(basePath, Arc.Entries.Count, () => Arc.Entries[0].Name, () => {
        if ($"{Path.GetFileNameWithoutExtension(path)}/" == Arc.Entries[0].FullName) basePath = $"{Arc.Entries[0].FullName}{basePath}";
        if (!string.IsNullOrEmpty(basePath)) Root = $"{basePath}{(basePath.EndsWith("/") ? "" : "/")}";
        return this;
    });
}

/// <summary>
/// SevenZipFileSystem
/// </summary>
public class SevenZipFileSystem(FileSystem vfx, string path, string basePath) : FileSystem {
    readonly SevenZipArchive Arc = SevenZipArchive.Open(vfx.Open(path));
    string Root = string.Empty;

    public override IEnumerable<string> Glob(string path, string searchPattern) {
        var root = Path.Combine(Root, path); var skip = root.Length;
        var matcher = CreateMatcher(searchPattern);
        return [.. Arc.Entries.Where(x =>
        {
            var fn = x.Key;
            return fn.Length > skip && fn.StartsWith(root) && matcher(fn[skip..]);
        }).Select(x => x.Key[skip..])];
    }
    public override bool FileExists(string path) { path = Path.Combine(Root, path); return Arc.Entries.Any(x => x.Key == path); }
    public override (string path, long length) FileInfo(string path) { path = Path.Combine(Root, path); var x = Arc.Entries.FirstOrDefault(x => x.Key == path); return x != null ? (x.Key, x.Size) : (null, 0); }
    public override Stream Open(string path, string mode) { path = Path.Combine(Root, path); return Arc.Entries.First(x => x.Key == path).OpenEntryStream(); }
    public override FileSystem Next() => Next2(basePath, Arc.Entries.Count, () => Arc.Entries.First().Key, () => {
        if (!string.IsNullOrEmpty(basePath)) Root = $"{basePath}{Path.AltDirectorySeparatorChar}";
        return this;
    });
}

#endregion

File: Python/openstk/vfx/vfx.py

from __future__ import annotations
import os, io, re, pathlib
from zipfile import ZipFile

#region FileSystem

# FileSystem
class FileSystem:
    def glob(self, path: str, searchPattern: str) -> list[str]: pass
    def fileExists(self, path: str) -> bool: pass
    def fileInfo(self, path: str) -> tuple[str, int]: pass
    def open(self, path: str, mode: str = None) -> object: pass
    def next(self) -> object: return self

    def findPaths(self, path: str, searchPattern: str) -> str:
        if (expandStartIdx := searchPattern.find('(')) != -1 and \
            (expandMidIdx := searchPattern.find(':', expandStartIdx)) != -1 and \
            (expandEndIdx := searchPattern.find(')', expandMidIdx)) != -1 and \
            expandStartIdx < expandEndIdx:
            for expand in searchPattern[expandStartIdx + 1: expandEndIdx].split(':'):
                for found in self.findPaths(path, searchPattern[:expandStartIdx] + expand + searchPattern[expandEndIdx+1:]): yield found
            return
        for path in self.glob(path, searchPattern): yield path

    def advance(self, basePath: str, path: str) -> object:
        match os.path.splitext(path)[1].lower():
            case '.zip': return ZipFileSystem(self, path, basePath)
            # case '.7z': return SevenZipFileSystem(self, path, basePath)
            # case '.iso' | '.bin' | '.cue': return DiscFileSystem(self, path, basePath)
            # case '.n64' | '.v64' | '.z64': return N64FileSystem(self, path, basePath)
            # case '.3ds': return X3dsFileSystem(self, path, basePath)
            case _: return None

    def next2(self, basePath: str, count: int, firstFunc: callable, elseFunc: callable) -> object:
        if count == 0: return self
        first = firstFunc(); firstLower = first.lower()
        if count == 1 or firstLower.endswith('.bin') or firstLower.endswith('.cue'): return self.advance(basePath, first) or self
        return elseFunc()

    @staticmethod
    def createMatcher(searchPattern: str) -> callable:
        if not searchPattern: return lambda x: True
        wildcardCount = searchPattern.count('*')
        if wildcardCount <= 0: return lambda x: x.casefold() == searchPattern.casefold()
        elif wildcardCount == 1:
            newPattern = searchPattern.replace('*', '')
            if searchPattern.startswith('*'): return lambda x: x.casefold().endswith(newPattern)
            elif searchPattern.endswith('*'): return lambda x: x.casefold().startswith(newPattern)
        regexPattern = f'^{re.escape(searchPattern).replace('\\*', '.*')}$'
        def lambdaX(x: str):
            try: return re.match(x, regexPattern)
            except: return False
        return lambdaX
    
# tag::AggregateFileSystem[]
class AggregateFileSystem(FileSystem):
    def __init__(self, aggreate: list[FileSystem]): self.aggreate = aggreate
    def glob(self, path: str, searchPattern: str) -> list[str]: return [y for z in [x.glob(path, searchPattern) for x in self.aggreate] for y in z]
    def fileExists(self, path: str) -> bool: return any(x.fileExists(path) for x in self.aggreate)
    def fileInfo(self, path: str) -> tuple[str, int]: return next(iter([x.fileInfo(path) for x in self.aggreate if x]))
    def open(self, path: str, mode: str = None) -> object: return next(iter([x.open(path, mode) for x in self.aggreate if x]))
# end::AggregateFileSystem[]

# tag::VirtualFileSystem[]
class VirtualFileSystem(FileSystem):
    def __init__(self, virtuals: dict[str, object]): self.virtuals = virtuals
    def glob(self, path: str, searchPattern: str) -> list[str]:
        matcher = FileSystem.createMatcher(searchPattern)
        return [x for x in self.virtuals.keys() if matcher(x)]
    def fileExists(self, path: str) -> bool: return path in self.virtuals
    def fileInfo(self, path: str) -> tuple[str, int]: return (path, len(self.virtuals[path]) if self.virtuals[path] else 0) if path in self.virtuals else (None, 0)
    def open(self, path: str, mode: str = None) -> object: return self.virtuals[path] or io.BytesIO() if path in self.virtuals else None
# end::VirtualFileSystem[]

# tag::DirectoryFileSystem[]
class DirectoryFileSystem(FileSystem):
    def __init__(self, baseRoot: str, basePath: str): self.baseRoot = baseRoot; self.basePath = basePath; self.root = baseRoot; self.skip = len(baseRoot) + 1
    def glob(self, path: str, searchPattern: str) -> list[str]:
        g = pathlib.Path(os.path.join(self.root, path)).glob(searchPattern if searchPattern else '**/*')
        return [str(x)[self.skip:] for x in g if x.is_file()]
    def fileExists(self, path: str) -> bool: return os.path.exists(os.path.join(self.root, path))
    def fileInfo(self, path: str) -> tuple[str, int]: return (path, os.stat(path).st_size) if os.path.exists(path := os.path.join(self.root, path)) else (None, 0)
    def open(self, path: str, mode: str = None) -> object: return open(os.path.join(self.root, path), mode or 'rb')
    def next(self) -> FileSystem:
        if os.path.isfile(self.root) or '*' in os.path.basename(self.root):
            self.root = os.path.dirname(self.root); self.skip = len(self.root) + 1
            return self #self.advance(self.basePath, ).next() or self
        @staticmethod
        def _lambdaX(self):
            if self.basePath: self.root = os.path.join(self.baseRoot, self.basePath); self.skip = len(self.root) + 1
            return self
        return self.next2(self.basePath, -1, lambda: next(iter(os.listdir(self.root)), None), lambda: _lambdaX(self))
# end::DirectoryFileSystem[]

#endregion

#region FileSystem : Network

# tag::NetworkFileSystem[]
class NetworkFileSystem(FileSystem):
    def __init__(self, uri: str): self.uri = uri
    def glob(self, path: str, searchPattern: str) -> list[str]: raise NotImplementedError()
    def fileExists(self, path: str) -> bool: raise NotImplementedError()
    def fileInfo(self, path: str) -> tuple[str, int]: raise NotImplementedError()
    def open(self, path: str, mode: str = None) -> object: raise NotImplementedError()
# end::NetworkFileSystem[]

#endregion

#region FileSystem : Archive

# tag::ZipFileSystem[]
class ZipFileSystem(FileSystem):
    def __init__(self, vfx: FileSystem, path: str, basePath: str): self.basePath = basePath; self.root = ''; self.zip = ZipFile(path)
    def glob(self, path: str, searchPattern: str) -> list[str]:
        root = os.path.join(self.root, path); skip = len(root); return []
    def fileExists(self, path: str) -> bool: return self.zip.read(os.path.join(self.root, path)) != None
    def fileInfo(self, path: str) -> tuple[str, int]: x = self.zip.read(os.path.join(self.root, path)); return (x.name, x.length) if x else (None, 0)
    def open(self, path: str, mode: str = None) -> object: return self.zip.read(os.path.join(self.root, path))
    def _lambdaX(self):
        if self.basePath: self.root = f'{path}/'
        return self
    def next(self) -> object: return self.next2(self.basePath, self.zip.count, lambda: self.zip.one, _lambdaX)
# end::ZipFileSystem[]

#endregion
